crates/sui-core/src/authority/authority_per_epoch_store.rs

```diff
@@ -2589,8 +2595,23 @@ impl AuthorityPerEpochStore {
             } else {
                 consensus_commit_info.round
             };
+
+            let mut checkpoint_roots: Vec<TransactionKey> = Vec::with_capacity(roots.len() + 1);
+
+            if let Some(consensus_commit_prologue_root) = consensus_commit_prologue_root {
+                if self
+                    .protocol_config()
+                    .prepose_consensus_commit_prologue_in_checkpoints()
+                {
+                    // Put consensus commit prologue root at the beginning of the checkpoint roots.
+                    checkpoint_roots.push(consensus_commit_prologue_root);
+                } else {
+                    roots.insert(consensus_commit_prologue_root);
+                }
+            }
```

Comment:
In theory, this part does not necessarily need to be guarded by the feature flag, because the causal sort in checkpoint builder is supposed to be input agnostic. However, I'm scared that if the order here somehow affects the causal sort order in checkpoint builder, it would cause fork when upgrading. So I still guarded preposing the consensus commit prologue here using the feature flag to be cautious.